### PR TITLE
Delete TensorBoardLogger experiment before spawning the processes.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @tchaton @awaelchli @edward-io

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 1]
```diff
@@ -415,3 +417,13 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    @staticmethod
+    def _clean_logger(*args: Any) -> None:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:422]: wasn't the argument that by moving the `_log_hyperparameters` we wouldn't need this?

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:422]: I thought so, but it still hanged.

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:422]: Reverted the change.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 2]
```diff
@@ -172,6 +173,7 @@ def spawn(self, function: Callable, *args: Any, return_result: bool = True, **kw
         Return:
             The output of the function of process 0.
         """
+        self._clean_logger(*args)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:176]: if this fix works, I propose to do it in the `start_x` methods above, before calling spawn().

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py]: Done !

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 3]
```diff
@@ -415,3 +419,11 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    @staticmethod
+    def _clean_logger(trainer: "pl.Trainer") -> None:
+        loggers = trainer._logger_iterable if isinstance(trainer.logger, LoggerCollection) else [trainer.logger]
+        for logger in loggers:
+            if isinstance(logger, TensorBoardLogger) and logger._experiment is not None:
+                rank_zero_warn("When using `ddp_spawn`, the Tensorboard experiment should be `None`.")
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:428]:
```suggestion
                rank_zero_warn(
                    "When using `ddp_spawn`, the `TensorBoardLogger` experiment should be `None`. Setting it to `None`."
                )
```

**Review by rohitgr7 — APPROVED** [review]
is it possible to add a test?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 4]
```diff
@@ -415,3 +419,11 @@ def teardown(self) -> None:
...
+                logger._experiment = None
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:429]: This doesn't guarantee that garbage collection will cleanup the writer instance (where I expect the spawn queue gets closed)

Should we del instead?

```suggestion
                del logger._experiment
                logger._experiment = None
```

or

```suggestion
                logger._experiment = None
                gc.collect()
```

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:428]: Done !

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:429]: Done !

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 5]
```diff
@@ -415,3 +420,15 @@ def teardown(self) -> None:
...
+    def _clean_logger(trainer: "pl.Trainer") -> None:
+        loggers = trainer.logger._logger_iterable if isinstance(trainer.logger, LoggerCollection) else [trainer.logger]
+        for logger in loggers:
+            if isinstance(logger, TensorBoardLogger) and logger._experiment is not None:
+                rank_zero_warn(
+                    "When using `ddp_spawn`, the `TensorBoardLogger` experiment should be `None`. Setting it to `None`."
+                )
+                del logger._experiment
+                logger._experiment = None
+                gc.collect()
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:434]: adding collect here is overkill. and also a source of slowness

if the reference count for the object in `logger._experiment` is 0 after deletion, python will instantly `__del__` it.

```suggestion
```

And if the reference count is >0, neither `del` nor `gc.collect()` will help

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:434]: See answer `2)` here: https://stackoverflow.com/a/6772907

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 6]
```diff
@@ -415,3 +419,16 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    @staticmethod
+    def _clean_logger(trainer: "pl.Trainer") -> None:
+        loggers = trainer.logger._logger_iterable if isinstance(trainer.logger, LoggerCollection) else [trainer.logger]
+        for logger in loggers:
+            if isinstance(logger, TensorBoardLogger) and logger._experiment is not None:
+                rank_zero_warn(
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:428]: @tchaton Is it necessary to have this warning? This info won't be necessary for majority of users. 

And this warning pops up every time a user trains with `ddp_spawn`, `ddp_spawn_sharded` and `tpu_spawn` as TensorBoard logger is the default.

**Pull request merged**